pennylane/data/__init__.py

```diff
@@ -24,7 +23,177 @@
+
+
+Creating a Dataset
+~~~~~~~~~~~~~~~~~~
+
+To create a new dataset in-memory, initialize a new ``Dataset`` with the desired attributes:
+
+    >>> hamiltonian = qml.Hamiltonian([1., 1.], [qml.PauliZ(wires=0), qml.PauliZ(wires=1)])
+    >>> eigvals, eigvecs = np.linalg.eigh(qml.matrix(hamiltonian))
+    >>> dataset = qml.data.Dataset(
+            hamiltonian = hamiltonian,
+            eigen = {"eigvals": eigvals, "eigvecs": eigvecs})
+    >>> dataset.hamiltonian
+    <Hamiltonian: terms=2, wires=[0, 1]>
+    >>> dataset.eigen
+    {'eigvals': array([-2.,  0.,  0.,  2.]),
+    'eigvecs': array([[0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j],
+       [0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 0.+0.j, 1.+0.j, 0.+0.j],
+       [1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j]])}
+
+Attributes can also be assigned to the instance after creation:
+
+    >>> dataset.ground_state = np.transpose(eigvecs)[np.argmin(eigvals)]
+    >>> dataset.ground_state
+    array([0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j])
+
+
+Reading and Writing Datasets
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Datasets can be saved to disk for later use. Datasets use the HDF5 format for serialization,
+which uses the '.h5' file extension.
+
+To save a dataset, use the :meth:`Dataset.write()` method:
+
+    >>> my_dataset = Dataset(...)
+    >>> my_dataset.write("~/datasets/my_dataset.h5")
+
+To open a dataset from a file, use :meth:`Dataset.open()` class method:
+
+    >>> my_dataset = Dataset.open("~/datasets/my_dataset.h5", mode="r")
+
+The `mode` argument follow the standard library convention - 'r' for reading, 'w-' and `w` for create and overwrite,
+and 'a' for editing. ``open()`` can be used to create a new dataset directly on disk:
+
+    >>> new_dataset = Dataset.open("~/datasets/new_datasets.h5", mode="w")
+
+By default, any changes made to an opened dataset will be committed directly to the file, which will fail
+if the file is opened read-only. The `"copy"` mode can be used to load the dataset into memory and detach
+it from the file:
+
+    >>> my_dataset = Dataset.open("~/dataset/my_dataset/h5", mode="copy")
+    >>> my_dataset.new_attribute = "abc"
+
+
+Attribute Metadata
+~~~~~~~~~~~~~~~~~~
+
+Dataset attributes can also contain additional metadata, such as docstrings. The :func:`qml.data.attribute`
+function can be used to attach metadata on assignment or initialization.
+
+    >>> hamiltonian = qml.Hamiltonian([1., 1.], [qml.PauliZ(wires=0), qml.PauliZ(wires=1)])
+    >>> eigvals, eigvecs = np.linalg.eigh(qml.matrix(hamiltonian))
+    >>> dataset = qml.data.Dataset(hamiltonian = qml.data.attribute(
+            hamiltonian,
+            doc="The hamiltonian of the system"))
+    >>> dataset.eigen = attribute(
+            {"eigvals": eigvals, "eigvecs": eigvecs},
+            doc="Eigenvalues and eigenvectors of the hamiltonain")
+
+This metadata can then be accessed using the :meth:`Dataset.attr_info` mapping:
+
+    >>> dataset.attr_info["hamiltonian"]["doc"]
```

Comment:
```suggestion
    >>> dataset.attr_info["eigen"]["doc"]
```
Assuming this is linked to above code example.